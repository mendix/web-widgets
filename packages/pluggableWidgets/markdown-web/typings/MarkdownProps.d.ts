/**
 * This file was generated from Markdown.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix Widgets Framework Team
 */
import { ActionValue, EditableValue } from "mendix";

export type ReadOnlyStyleEnum = "bordered" | "text";

export interface MarkdownContainerProps {
    name: string;
    tabIndex?: number;
    id: string;
    stringAttribute: EditableValue<string>;
    spellcheck: boolean;
    showFooter: boolean;
    readOnlyStyle: ReadOnlyStyleEnum;
    onChangeEvent?: ActionValue;
    onEnterEvent?: ActionValue;
    onLeaveEvent?: ActionValue;
}

export interface MarkdownPreviewProps {
    readOnly: boolean;
    stringAttribute: string;
    spellcheck: boolean;
    showFooter: boolean;
    readOnlyStyle: ReadOnlyStyleEnum;
    onChangeEvent: {} | null;
    onEnterEvent: {} | null;
    onLeaveEvent: {} | null;
}
