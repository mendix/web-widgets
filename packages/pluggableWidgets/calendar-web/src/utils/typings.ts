import { ObjectItem } from "mendix";
import { CalendarProps as ReactCalendarProps } from "react-big-calendar";
import { withDragAndDropProps } from "react-big-calendar/lib/addons/dragAndDrop";

export interface CalendarEvent {
    title: string;
    start: Date;
    end: Date;
    allDay: boolean;
    color?: string;
    item: ObjectItem;
}

export type EventDropOrResize = {
    event: CalendarEvent;
    start: Date;
    end: Date;
};

export interface DragAndDropCalendarProps<TEvent extends object = Event, TResource extends object = object>
    extends ReactCalendarProps<TEvent, TResource>,
        withDragAndDropProps<TEvent, TResource> {}
