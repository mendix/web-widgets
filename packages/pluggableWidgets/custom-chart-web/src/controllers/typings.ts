import { ActionValue, EditableValue } from "mendix";

export interface ControllerProps {
    configurationOptions: string;
    dataAttribute?: EditableValue<string>;
    dataStatic: string;
    eventDataAttribute?: EditableValue<string>;
    layoutAttribute?: EditableValue<string>;
    layoutStatic: string;
    onClick?: ActionValue;
    sampleData: string;
    sampleLayout: string;
}
