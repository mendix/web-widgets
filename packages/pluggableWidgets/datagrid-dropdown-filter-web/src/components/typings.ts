import { RefFilterStore } from "@mendix/widget-plugin-dropdown-filter/stores/RefFilterStore";
import { StaticSelectFilterStore } from "@mendix/widget-plugin-dropdown-filter/stores/StaticSelectFilterStore";

export interface EnumFilterAPI {
    filterStore: StaticSelectFilterStore;
    parentChannelName?: string;
}

export interface RefFilterAPI {
    filterStore: RefFilterStore;
    parentChannelName?: string;
}
