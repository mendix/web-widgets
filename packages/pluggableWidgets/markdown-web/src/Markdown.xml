<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.web.markdown.Markdown" needsEntityContext="true" pluginWidget="true" offlineCapable="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../../node_modules/mendix/custom_widget.xsd">
    <name>Markdown viewer</name>
    <description />
    <studioProCategory>Input elements</studioProCategory>
    <studioCategory>Input Elements</studioCategory>
    <helpUrl>https://docs.mendix.com/appstore/widgets/markdown</helpUrl>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="Data source">
                <property key="stringAttribute" type="attribute">
                    <caption>Value attribute</caption>
                    <description>The attribute used for the content of the text editor, recommendation is to use an unlimited string data type.</description>
                    <attributeTypes>
                        <attributeType name="String" />
                    </attributeTypes>
                </property>
            </propertyGroup>
            <!-- SYSTEM PROPS -->
            <propertyGroup caption="Label">
                <systemProperty key="Label" />
            </propertyGroup>

            <propertyGroup caption="Conditional visibility">
                <systemProperty key="Visibility" />
            </propertyGroup>
            <!-- END SYSTEM PROPS -->
        </propertyGroup>
    </properties>
</widget>
