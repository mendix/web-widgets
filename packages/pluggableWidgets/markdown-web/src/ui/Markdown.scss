$spacing-smaller: 4px;
$label-info-color: #fff;
$gray-primary: #ced0d3;
$gray-lighter: #f8f8f8;
$table-background: #f2f2f2;

.widget-markdown {
    flex-grow: 1;
    min-height: 300px;
    border: 1px solid var(--gray-primary);
    border-radius: var(--spacing-smaller, $spacing-smaller);
    display: flex;
    justify-content: flex-start;
    align-items: center;
    flex-direction: column;

    &-topbar {
        width: 100%;
        height: 40px;
        border-top-left-radius: var(--spacing-smaller, $spacing-smaller);
        border-top-right-radius: var(--spacing-smaller, $spacing-smaller);
        background-color: var(--gray-lighter);

        &-preview-button {
            background-color: transparent;
            border: none;
            margin: 0;
            padding: 0 15px;
            height: 100%;
            border-top-left-radius: var(--spacing-smaller, $spacing-smaller);
            border-top-right-radius: var(--spacing-smaller, $spacing-smaller);
            cursor: none;
            &.active {
                border-left: 1px solid var(--gray-primary);
                border-right: 1px solid var(--gray-primary);
                background-color: var(--label-info-color, $label-info-color);
            }
        }
    }

    &-content {
        flex: 1;
        width: 100%;
        display: flex;
        justify-content: center;
        align-items: center;
        background-color: var(--label-info-color, $label-info-color);
        padding: 5px;
        border-radius: var(--spacing-smaller, $spacing-smaller);

        &-preview {
            padding: 10px;
            background-color: #fff;
            min-height: 200px;
            max-width: 90% table {
                width: 100%;
                border-collapse: collapse;
                margin-bottom: 10px;
            }

            th,
            td {
                border: 1px solid #ccc;
                padding: 8px;
                text-align: left;
            }

            th {
                background-color: $table-background;
            }

            img {
                max-width: 35%;
            }
        }
    }

    &-footer {
        width: 100%;
        height: 30px;
        border-top: 1px solid var(--gray-primary);
        display: flex;
        justify-content: flex-start;
        align-items: center;
        gap: 15px;
        padding: 0 10px;
        font-size: var(--font-size-small, 12px);
    }

    &-icon {
        display: flex;
    }
}
