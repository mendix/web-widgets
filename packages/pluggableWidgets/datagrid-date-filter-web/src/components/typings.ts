import { Date_InputFilterInterface } from "@mendix/widget-plugin-filtering/typings/InputFilterInterface";

export interface DateFilterProps {
    filterStore: Date_InputFilterInterface;
    parentChannelName?: string;
}
