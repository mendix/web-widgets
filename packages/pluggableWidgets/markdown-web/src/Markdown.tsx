import classNames from "classnames";
import MarkdownIt from "markdown-it";
import { ReactElement, createElement, useEffect, useRef } from "react";
import { MarkdownContainerProps } from "../typings/MarkdownProps";
import "./ui/Markdown.scss";
const mdParser = new MarkdownIt("default", {
    typographer: true,
    linkify: true
});

export default function Markdown(props: MarkdownContainerProps): ReactElement {
    const { stringAttribute } = props;
    const previewRef = useRef<HTMLDivElement>(null);
    useEffect(() => {
        if (previewRef.current) {
            previewRef.current.innerHTML = mdParser.render(stringAttribute.value ?? "");
        }
    }, [stringAttribute?.value, stringAttribute?.status]);

    return stringAttribute?.status === "available" ? (
        <div className={classNames("widget-markdown")} ref={previewRef}>
            {stringAttribute?.value}
        </div>
    ) : (
        <div className="mx-progress"></div>
    );
}
