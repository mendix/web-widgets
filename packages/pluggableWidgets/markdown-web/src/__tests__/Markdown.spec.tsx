import { EditableValueBuilder } from "@mendix/widget-plugin-test-utils";
import "@testing-library/jest-dom";
import { render } from "@testing-library/react";
import { createElement } from "react";
import { MarkdownContainerProps } from "typings/MarkdownProps";
import Markdown from "../Markdown";

describe("Markdown widget", () => {
    let defaultProps: MarkdownContainerProps;
    beforeEach(() => {
        defaultProps = {
            name: "Markdown1",
            id: "Markdown1",
            stringAttribute: new EditableValueBuilder<string>().withValue("Markdown viewer default value").build()
        };
    });

    it("renders markdown widget", () => {
        const component = render(<Markdown {...defaultProps} />);
        const markdownElement = component.container.querySelector(".widget-markdown");
        expect(markdownElement).toBeInTheDocument();
        expect(component.container).toMatchSnapshot();
    });

    it("does not render markdown widget when stringAttribute is unavailable", () => {
        defaultProps.stringAttribute = new EditableValueBuilder<string>().isUnavailable().build();
        const component = render(<Markdown {...defaultProps} />);
        const progressElement = component.container.querySelector(".mx-progress");
        expect(progressElement).toBeInTheDocument();
        expect(component.container).toMatchSnapshot();
    });
});
