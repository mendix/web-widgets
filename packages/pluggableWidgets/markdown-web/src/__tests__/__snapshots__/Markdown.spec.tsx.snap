// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markdown widget does not render markdown widget when stringAttribute is unavailable 1`] = `
<div>
  <div
    class="mx-progress"
  />
</div>
`;

exports[`Markdown widget renders markdown widget 1`] = `
<div>
  <div
    class="widget-markdown"
  >
    Markdown viewer default value
  </div>
</div>
`;
